awx/main/models/__init__.py

```diff
@@ -122,18 +122,22 @@ def user_is_system_auditor(user):
 
 @user_is_system_auditor.setter
 def user_is_system_auditor(user, tf):
-    if user.id:
-        if tf:
-            role = Role.singleton('system_auditor')
-            # must check if member to not duplicate activity stream
-            if user not in role.members.all():
-                role.members.add(user)
-            user._is_system_auditor = True
-        else:
-            role = Role.singleton('system_auditor')
-            if user in role.members.all():
-                role.members.remove(user)
-            user._is_system_auditor = False
+    if not user.id:
+        # If the user doesn't have a primary key yet (i.e., this is the *first*
+        # time they've logged in, and we've just created the new User in this
+        # request), we need one to set up the system auditor role
+        user.save()
```

Comment:
The getter would also probably need modification, `user_is_system_auditor` above this.

Reply:
I dug through this code and I don't see any cases where the login _gets_ the attribute before setting it.

This line sets `._is_system_auditor`:

https://github.com/ansible/awx/pull/4396/files#diff-b00f0ac10475072c791a4f7d06782447L131

...and if we do a `.is_system_auditor` later, we'll hit this line:

https://github.com/ansible/awx/pull/4396/files#diff-b00f0ac10475072c791a4f7d06782447R120

So the only way you should enter this block:

https://github.com/ansible/awx/pull/4396/files#diff-b00f0ac10475072c791a4f7d06782447R118

...is if you do `.is_system_auditor` on a User *before* it's saved.  I can't find any login code paths where that's the case, and it seems like that's _always_ been the behavior (given the comment).

Reply:
It looks to me like it does not need to be modified, since it already has a defensive check for the case where the user object hasn't been saved yet.

Reply:
TFW you see someone has already responded more extensively than oneself after you hit 'comment' :man_facepalming:

Reply:
🤷‍♂